File: libs/go/limits.go
package stonecharts

import (
	"fmt"
	"unicode/utf8"
)

const (
	MaxSpecBytes       = 1000000
	MaxSeries          = 50
	MaxPointsPerSeries = 10000
	MaxTotalPoints     = 50000
	MaxLabelLength     = 512
	MaxSVGBytes        = 5000000
)

type ResourceLimitError struct {
	Code     string
	Path     string
	Limit    int
	Received int
}

func (e *ResourceLimitError) Error() string {
	return fmt.Sprintf("%s: %s: limit %d exceeded, received %d", e.Code, e.Path, e.Limit, e.Received)
}

func checkLabelLimit(v interface{}, path string) error {
	if s, ok := v.(string); ok && utf8.RuneCountInString(s) > MaxLabelLength {
		return &ResourceLimitError{Code: "LIMIT.LABEL_LENGTH", Path: path, Limit: MaxLabelLength, Received: utf8.RuneCountInString(s)}
	}
	return nil
}

func enforceSpecLimits(spec interface{}, rawSizeHint int) error {
	if rawSizeHint > MaxSpecBytes {
		return &ResourceLimitError{Code: "LIMIT.SPEC_BYTES", Path: "$", Limit: MaxSpecBytes, Received: rawSizeHint}
	}
	d, ok := spec.(map[string]interface{})
	if !ok {
		return nil
	}
	for _, key := range []string{"id", "title", "subtitle"} {
		if err := checkLabelLimit(d[key], "$."+key); err != nil {
			return err
		}
	}
	for _, axisName := range []string{"xAxis", "yAxis", "secondaryYAxis"} {
		axis, ok := d[axisName].(map[string]interface{})
		if !ok {
			continue
		}
		if err := checkLabelLimit(axis["title"], "$."+axisName+".title"); err != nil {
			return err
		}
		if categories, ok := axis["categories"].([]interface{}); ok {
			for i, category := range categories {
				if err := checkLabelLimit(category, "$."+axisName+".categories["+itoa(i)+"]"); err != nil {
					return err
				}
			}
		}
	}
	series, ok := d["series"].([]interface{})
	if !ok {
		return nil
	}
	if len(series) > MaxSeries {
		return &ResourceLimitError{Code: "LIMIT.SERIES_COUNT", Path: "$.series", Limit: MaxSeries, Received: len(series)}
	}
	totalPoints := 0
	for seriesIndex, item := range series {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		if err := checkLabelLimit(m["name"], "$.series["+itoa(seriesIndex)+"].name"); err != nil {
			return err
		}
		data, ok := m["data"].([]interface{})
		if !ok {
			continue
		}
		if len(data) > MaxPointsPerSeries {
			return &ResourceLimitError{Code: "LIMIT.POINTS_PER_SERIES", Path: "$.series[" + itoa(seriesIndex) + "].data", Limit: MaxPointsPerSeries, Received: len(data)}
		}
		totalPoints += len(data)
		if totalPoints > MaxTotalPoints {
			return &ResourceLimitError{Code: "LIMIT.TOTAL_POINTS", Path: "$.series[*].data", Limit: MaxTotalPoints, Received: totalPoints}
		}
	}
	return nil
}

func enforceSVGLimit(svg string) error {
	size := len([]byte(svg))
	if size > MaxSVGBytes {
		return &ResourceLimitError{Code: "LIMIT.SVG_BYTES", Path: "$.svg", Limit: MaxSVGBytes, Received: size}
	}
	return nil
}

File: libs/python/stonecharts/limits.py
"""Resource limits for certified StoneCharts renderers."""

from __future__ import annotations

from typing import Any

MAX_SPEC_BYTES = 1_000_000
MAX_SERIES = 50
MAX_POINTS_PER_SERIES = 10_000
MAX_TOTAL_POINTS = 50_000
MAX_LABEL_LENGTH = 512
MAX_SVG_BYTES = 5_000_000


class ResourceLimitError(ValueError):
    """Raised when a spec or output exceeds a documented resource limit."""

    def __init__(self, code: str, path: str, limit: int, received: int):
        self.code = code
        self.path = path
        self.limit = limit
        self.received = received
        super().__init__(f"{code}: {path}: limit {limit} exceeded, received {received}")


def _check_label(value: Any, path: str) -> None:
    if isinstance(value, str) and len(value) > MAX_LABEL_LENGTH:
        raise ResourceLimitError("LIMIT.LABEL_LENGTH", path, MAX_LABEL_LENGTH, len(value))


def enforce_spec_limits(spec: Any, *, raw_size_hint: int | None = None) -> None:
    if raw_size_hint is not None and raw_size_hint > MAX_SPEC_BYTES:
        raise ResourceLimitError("LIMIT.SPEC_BYTES", "$", MAX_SPEC_BYTES, raw_size_hint)
    if not isinstance(spec, dict):
        return
    for key in ("id", "title", "subtitle"):
        _check_label(spec.get(key), f"$.{key}")
    for axis_name in ("xAxis", "yAxis", "secondaryYAxis"):
        axis = spec.get(axis_name)
        if isinstance(axis, dict):
            _check_label(axis.get("title"), f"$.{axis_name}.title")
            categories = axis.get("categories")
            if isinstance(categories, list):
                for index, category in enumerate(categories):
                    _check_label(category, f"$.{axis_name}.categories[{index}]")
    series = spec.get("series")
    if not isinstance(series, list):
        return
    if len(series) > MAX_SERIES:
        raise ResourceLimitError("LIMIT.SERIES_COUNT", "$.series", MAX_SERIES, len(series))
    total_points = 0
    for series_index, item in enumerate(series):
        if not isinstance(item, dict):
            continue
        _check_label(item.get("name"), f"$.series[{series_index}].name")
        data = item.get("data")
        if not isinstance(data, list):
            continue
        point_count = len(data)
        if point_count > MAX_POINTS_PER_SERIES:
            raise ResourceLimitError(
                "LIMIT.POINTS_PER_SERIES",
                f"$.series[{series_index}].data",
                MAX_POINTS_PER_SERIES,
                point_count,
            )
        total_points += point_count
        if total_points > MAX_TOTAL_POINTS:
            raise ResourceLimitError("LIMIT.TOTAL_POINTS", "$.series[*].data", MAX_TOTAL_POINTS, total_points)


def enforce_svg_limit(svg: str) -> None:
    size = len(svg.encode("utf-8"))
    if size > MAX_SVG_BYTES:
        raise ResourceLimitError("LIMIT.SVG_BYTES", "$.svg", MAX_SVG_BYTES, size)
